File: gremlin-util/src/main/java/org/apache/tinkerpop/gremlin/util/ser/binary/ResponseMessageSerializer.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.tinkerpop.gremlin.util.ser.binary;

import io.netty.buffer.ByteBuf;
import org.apache.tinkerpop.gremlin.util.ser.NettyBufferFactory;
import org.apache.tinkerpop.gremlin.util.message.ResponseMessage;
import org.apache.tinkerpop.gremlin.util.message.ResponseResult;
import org.apache.tinkerpop.gremlin.util.message.ResponseStatus;
import org.apache.tinkerpop.gremlin.util.message.ResponseStatusCode;
import org.apache.tinkerpop.gremlin.util.ser.SerializationException;
import org.apache.tinkerpop.gremlin.structure.io.Buffer;
import org.apache.tinkerpop.gremlin.structure.io.binary.GraphBinaryReader;
import org.apache.tinkerpop.gremlin.structure.io.binary.GraphBinaryWriter;

import java.io.IOException;
import java.util.Map;
import java.util.UUID;

public class ResponseMessageSerializer {
    private static final NettyBufferFactory bufferFactory = new NettyBufferFactory();

    public ResponseMessage readValue(final ByteBuf byteBuf, final GraphBinaryReader context) throws SerializationException {
        // Wrap netty's buffer
        final Buffer buffer = bufferFactory.create(byteBuf);
        final int version = buffer.readByte() & 0xff;

        if (version >>> 7 != 1) {
            // This is an indication that the response buffer was incorrectly built
            // Or the buffer offsets are wrong
            throw new SerializationException("The most significant bit should be set according to the format");
        }

        try {
            return ResponseMessage.build(context.readValue(buffer, UUID.class, true))
                    .code(ResponseStatusCode.getFromValue(context.readValue(buffer, Integer.class, false)))
                    .statusMessage(context.readValue(buffer, String.class, true))
                    .statusAttributes(context.readValue(buffer, Map.class, false))
                    .responseMetaData(context.readValue(buffer, Map.class, false))
                    .result(context.read(buffer))
                    .create();
        } catch (IOException ex) {
            throw new SerializationException(ex);
        }
    }

    public void writeValue(final ResponseMessage value, final ByteBuf byteBuf, final GraphBinaryWriter context) throws SerializationException {
        // Wrap netty's buffer
        final Buffer buffer = bufferFactory.create(byteBuf);

        final ResponseResult result = value.getResult();
        final ResponseStatus status = value.getStatus();

        try {
            // Version
            buffer.writeByte(GraphBinaryWriter.VERSION_BYTE);
            // Nullable request id
            context.writeValue(value.getRequestId(), buffer, true);
            // Status code
            context.writeValue(status.getCode().getValue(), buffer, false);
            // Nullable status message
            context.writeValue(status.getMessage(), buffer, true);
            // Status attributes
            context.writeValue(status.getAttributes(), buffer, false);
            // Result meta
            context.writeValue(result.getMeta(), buffer, false);
            // Fully-qualified value
            context.write(result.getData(), buffer);
        } catch (IOException ex) {
            throw new SerializationException(ex);
        }
    }
}

File: gremlin-dotnet/src/Gremlin.Net/Structure/IO/GraphBinary4/ResponseMessageSerializer.cs
#region License

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#endregion

using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Gremlin.Net.Driver.Exceptions;
using Gremlin.Net.Driver.Messages;
using Gremlin.Net.Process.Traversal;

namespace Gremlin.Net.Structure.IO.GraphBinary4
{
    /// <summary>
    ///     Deserializes a <see cref="ResponseMessage{T}"/> from the GraphBinary 4.0 streaming format.
    /// </summary>
    public class ResponseMessageSerializer
    {
        /// <summary>
        ///     Reads a response message from the stream in the 4.0 format:
        ///     version byte + bulked flag + result data + marker + status footer.
        /// </summary>
        /// <param name="stream">The GraphBinary data to parse.</param>
        /// <param name="reader">A <see cref="GraphBinaryReader"/> that can be used to read nested values.</param>
        /// <param name="cancellationToken">The token to cancel the operation. The default value is None.</param>
        /// <returns>The read response message.</returns>
        public async Task<ResponseMessage<List<object>>> ReadValueAsync(MemoryStream stream,
            GraphBinaryReader reader, CancellationToken cancellationToken = default)
        {
            // 1. Version byte — validate MSB is set
            var version = await stream.ReadByteAsync(cancellationToken).ConfigureAwait(false) & 0xFF;
            if (version >> 7 != 1)
            {
                throw new IOException("The most significant bit should be set according to the format");
            }

            // 2. Bulked flag
            var bulkedByte = await stream.ReadByteAsync(cancellationToken).ConfigureAwait(false);
            var bulked = bulkedByte == 0x01;

            // 3. Read result data until Marker
            var results = new List<object>();
            while (true)
            {
                var value = await reader.ReadAsync(stream, cancellationToken).ConfigureAwait(false);

                if (value is Marker)
                {
                    break;
                }

                if (bulked)
                {
                    // Read the fully-qualified Long bulk count following the value
                    var bulkCount = (long)(await reader.ReadAsync(stream, cancellationToken)
                        .ConfigureAwait(false))!;
                    results.Add(new Traverser(value, bulkCount));
                }
                else
                {
                    results.Add(value!);
                }
            }

            // 4. Status footer: status code (Int value)
            var statusCode = (int)await reader.ReadNonNullableValueAsync<int>(
                stream, cancellationToken).ConfigureAwait(false);

            // 5. Nullable status message (value_flag + String value)
            var statusMessage = (string?)await reader.ReadNullableValueAsync<string>(
                stream, cancellationToken).ConfigureAwait(false);

            // 6. Nullable exception (value_flag + String value)
            var exception = (string?)await reader.ReadNullableValueAsync<string>(
                stream, cancellationToken).ConfigureAwait(false);

            if (statusCode != 200)
            {
                throw new ResponseException(statusCode, statusMessage, exception);
            }

            return new ResponseMessage<List<object>>(bulked, results, statusCode,
                statusMessage, exception);
        }
    }
}
